lsp-mode.el

```diff
@@ -7563,13 +7596,13 @@ When ARG is t the lsp mode will start new language server even if
 there is language server which can handle current language. When
 ARG is nil current file will be opened in multi folder language
 server if there is such. When `lsp' is called with prefix
-argument ask the user to select which language server to start. "
+argument ask the user to select which language server to start."
   (interactive "P")
 
   (when (and lsp-auto-configure (not lsp--client-packages-required))
     (seq-do (lambda (package)
               ;; loading client is slow and `lsp' can be called repeatedly
-              (unless (featurep package) (require package nil t)))
```

Comment:
This is intended. why removing `unless` stuff?

Reply:
`require' already checks whether the feature is provided, meaning that we are doing redundant work. The `require' becomes a no-op if package is `featurep'.

Reply:
Well, there's a bug that reloading clients is slow, and the solution is to check `featurep` before requiring again.
Is there docs/notes saying that `require` will not load the package again?

If the file only contains function definition, it would fast to be reloaded since the function's not required to be reparse again, however, client packages contains `lsp-register-client`, which is run every time file is loaded, thus slow.
See https://github.com/emacs-lsp/lsp-mode/pull/1498

Reply:
"If feature FEATURE is not loaded, load it from FILENAME."

Also, try the following: (benchmark-run 100000 (require 'lsp-mode)). We don't even have 100k servers, and that completes in under a second.

Reply:
`featurep' is slightly faster: when called 100k times, it is in total 60ms faster. But for the max 100 servers we have, any speedups are negligible.

Reply:
> If feature FEATURE is not loaded, load it from FILENAME.

I'm not sure if that mean `if feature is loaded, not load it again`??
From the profiler in #1498, it said otherwise though.
If there's new behavior of `require`, did it happens in `Emacs 26`?

I think the slowness comes from code that's executed when file is loaded. Requiring `lsp-mode`, which doesn't execute `lsp` on top level, is not much different.

Reply:
Let's continue discussion here https://github.com/emacs-lsp/lsp-mode/pull/2208/files#r507208878.
I think we can get that PR merge first so it will be easier for you to separate your work

Reply:
It is only slow the first time it is run, because require isn't fast.

Here is how require works in Emacs:
There is a set of symbols, called the "features". To check whether a feature is provided (a feature just being a symbol), the `feature` function, implemented in C, checks whether that symbol is in that. A feature can be added to that set using the `provide` function, which is why all modules should end in (provide 'name).

Require first checks if some feature is provided, and, if not, tries to find the module that it needs to load. This is slow, since it has to check lots of files (the entire `load-path`, 327 directories on my system), but only has to be done once. So (require 'foo) might be somewhat slow the first time, but 'foo will then be provided, so subsequent requires are fast. The issue with lsp-mode was (is?) that some clients which aren't installed are required. This always fails, since those clients aren't in the `load-path`, meaning that those clients never get their features provided, so the `load-path` has to be traversed all the time, which is slow.

I hope that clears things up.